File: android/src/main/java/cn/thinkingdata/tdremoteconfig_flutter/TDRemoteConfigFlutterPlugin.java
/*
 * Copyright (C) 2025 ThinkingData
 */
package cn.thinkingdata.tdremoteconfig_flutter;

import android.content.Context;

import androidx.annotation.NonNull;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import cn.thinkingdata.remoteconfig.TDRemoteConfig;
import cn.thinkingdata.remoteconfig.TDRemoteConfigSettings;
import cn.thinkingdata.remoteconfig.core.TDObject;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

/**
 * @author liulongbing
 * @since 2025/6/25
 */
public class TDRemoteConfigFlutterPlugin implements FlutterPlugin, MethodChannel.MethodCallHandler {

    private MethodChannel channel;
    private Context mContext;
    private EventChannel.EventSink eventSink;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        mContext = binding.getApplicationContext();
        channel = new MethodChannel(binding.getBinaryMessenger(), "thinkingdata.cn/RemoteConfig");
        channel.setMethodCallHandler(this);
        EventChannel eventChannel = new EventChannel(binding.getBinaryMessenger(), "thinkingdata.cn/RemoteConfig/event");
        eventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object arguments, EventChannel.EventSink events) {
                eventSink = events;
            }

            @Override
            public void onCancel(Object arguments) {
                eventSink = null;
            }
        });
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        try {
            if ("enableLog".equals(call.method)) {
                boolean enableLog = ( boolean ) call.arguments;
                TDRemoteConfig.enableLog(enableLog);
                result.success(null);
            } else if ("initWithConfig".equals(call.method)) {
                TDRemoteConfigSettings settings = new TDRemoteConfigSettings();
                settings.appId = call.argument("appId");
                settings.serverUrl = call.argument("serverUrl");
                Integer mode = call.argument("mode");
                if (mode != null && mode == 1) {
                    settings.mode = TDRemoteConfigSettings.TDRemoteConfigMode.DEBUG;
                }
                Map<String, Object> fetchParams = call.argument("customFetchParams");
                settings.setCustomFetchParams(extractJSONObject(fetchParams));
                TDRemoteConfig.init(mContext, settings);
                result.success(null);
            } else if ("setDefaultValues".equals(call.method)) {
                Map<String, Object> defaultValues = call.argument("defaultValues");
                String appId = call.argument("appId");
                TDRemoteConfig.setDefaultValues(extractJSONObject(defaultValues), appId);
                result.success(null);
            } else if ("setDefaultValuesWithFile".equals(call.method)) {
                String filePath = call.argument("filePath");
                String appId = call.argument("appId");
                TDRemoteConfig.setDefaultValues(filePath, appId);
                result.success(null);
            } else if ("clearDefaultValues".equals(call.method)) {
                String appId = call.argument("appId");
                TDRemoteConfig.clearDefaultValues(appId);
                result.success(null);
            } else if ("setCustomFetchParams".equals(call.method)) {
                Map<String, Object> fetchParams = call.argument("fetchParams");
                String appId = call.argument("appId");
                String templateCode = call.argument("templateCode");
                TDRemoteConfig.setCustomFetchParams(extractJSONObject(fetchParams), appId, templateCode);
                result.success(null);
            } else if ("fetch".equals(call.method)) {
                String appId = call.argument("appId");
                String templateCode = call.argument("templateCode");
                TDRemoteConfig.fetch(appId, templateCode);
                result.success(null);
            } else if ("removeCustomFetchParam".equals(call.method)) {
                String appId = call.argument("appId");
                String key = call.argument("key");
                String templateCode = call.argument("templateCode");
                TDRemoteConfig.removeCustomFetchParam(key, appId, templateCode);
                result.success(null);
            } else if ("addClientParams".equals(call.method)) {
                Map<String, Object> clientParams = call.argument("clientParams");
                TDRemoteConfig.addClientParams(extractJSONObject(clientParams));
                result.success(null);
            } else if ("getClientParams".equals(call.method)) {
                JSONObject clientParams = TDRemoteConfig.getClientParams();
                result.success(jsonToMap(clientParams));
            } else if ("removeClientParam".equals(call.method)) {
                String key = call.argument("key");
                TDRemoteConfig.removeClientParam(key);
                result.success(null);
            } else if ("accumulateNum".equals(call.method)) {
                String key = call.argument("key");
                Double num = call.argument("num");
                if (num != null) {
                    TDRemoteConfig.accumulateNum(key, num);
                }
                result.success(null);
            } else if ("getClientParamValueForKey".equals(call.method)) {
                String key = call.argument("key");
                result.success(TDRemoteConfig.getClientParamValueForKey(key));
            } else if ("getConfigValue".equals(call.method)) {
                String appId = call.argument("appId");
                String templateCode = call.argument("templateCode");
                List<Map<String, Object>> dataList = call.argument("dataList");
                String type = call.argument("type");
                if (dataList != null) {
                    TDObject tdObject = TDRemoteConfig.getData(appId, templateCode);
                    for (Map<String, Object> stringObjectMap : dataList) {
                        if (stringObjectMap == null) continue;
                        Object key = stringObjectMap.get("key");
                        Object defaultValue = stringObjectMap.get("defaultValue");
                        if (key instanceof String) {
                            tdObject = tdObject.get(( String ) key, defaultValue);
                        } else if (key instanceof Integer) {
                            tdObject = tdObject.get(( Integer ) key, defaultValue);
                        }
                    }
                    if ("string".equals(type)) {
                        result.success(tdObject.stringValue());
                        return;
                    } else if ("int".equals(type)) {
                        result.success(tdObject.longValue());
                        return;
                    } else if ("double".equals(type)) {
                        result.success(tdObject.doubleValue());
                        return;
                    } else if ("bool".equals(type)) {
                        result.success(tdObject.booleanValue());
                        return;
                    } else if ("object".equals(type)) {
                        result.success(jsonToMap(tdObject.objectValue()));
                        return;
                    } else if ("array".equals(type)) {
                        result.success(jsonToList(tdObject.arrayValue()));
                        return;
                    }
                }
                result.success(null);
            } else if ("addConfigFetchListener".equals(call.method)) {
                TDRemoteConfig.addConfigFetchListener(new TDRemoteConfig.OnConfigFetchListener() {
                    @Override
                    public void onFetchSuccess(JSONObject jsonObject) {
                        if (eventSink != null) {
                            eventSink.success(jsonToMap(jsonObject));
                        }
                    }
                });
                result.success(null);
            } else {
                result.notImplemented();
            }
        } catch (Exception ignore) {
        }
    }

    private JSONObject extractJSONObject(Map<String, Object> properties) throws JSONException {
        JSONObject jsonObject = new JSONObject();
        if (properties != null) {
            for (String key : properties.keySet()) {
                Object value = properties.get(key);
                if (value instanceof Map<?, ?>) {
                    value = extractJSONObject(( Map<String, Object> ) value);
                } else if (value instanceof List) {
                    //value = new JSONArray(value);
                    value = new JSONArray(( List ) value);
                }
                jsonObject.put(key, value);
            }
        }
        return jsonObject;
    }

    public static Map<String, Object> jsonToMap(JSONObject jsonObject) {
        Map<String, Object> map = new HashMap<>();
        if (jsonObject == null) return map;
        Iterator<String> keys = jsonObject.keys();

        while (keys.hasNext()) {
            String key = keys.next();
            Object value = jsonObject.opt(key);

            // 递归处理嵌套的 JSONObject
            if (value instanceof JSONObject) {
                value = jsonToMap(( JSONObject ) value);
            }
            // 递归处理嵌套的 JSONArray
            else if (value instanceof JSONArray) {
                value = jsonToList(( JSONArray ) value);
            }

            map.put(key, value);
        }
        return map;
    }

    private static List<Object> jsonToList(JSONArray jsonArray) {
        List<Object> list = new ArrayList<>();
        if (jsonArray == null) return list;
        for (int i = 0; i < jsonArray.length(); i++) {
            Object value = jsonArray.opt(i);

            // 递归处理嵌套的 JSONObject
            if (value instanceof JSONObject) {
                list.add(jsonToMap(( JSONObject ) value));
            }
            // 递归处理嵌套的 JSONArray
            else if (value instanceof JSONArray) {
                list.add(jsonToList(( JSONArray ) value));
            } else {
                list.add(value);
            }
        }
        return list;
    }

}

File: ios/Classes/TDRemoteConfigFlutterPlugin.m
#import "TDRemoteConfigFlutterPlugin.h"
#import <TDRemoteConfig/TDRemoteConfig.h>

@interface TDRemoteConfigReceiver : NSObject
+ (void)setEventSink:(FlutterEventSink)eventSink;
@end

@implementation TDRemoteConfigReceiver

static FlutterEventSink sEventSink = nil;

+ (void)setEventSink:(FlutterEventSink)eventSink {
    sEventSink = [eventSink copy];
}

+ (void)fun:(NSNotification *)notification {
    NSDictionary *info = notification.userInfo[kTDRemoteConfigStrategyStatusMap];
    if (!info) {
        info = @{}; // 设置一个空字典作为默认值
    }
    dispatch_async(dispatch_get_main_queue(), ^{
        if(sEventSink){
            sEventSink(info);
        }
    });
}

@end

@implementation TDRemoteConfigFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"thinkingdata.cn/RemoteConfig"
            binaryMessenger:[registrar messenger]];
  TDRemoteConfigFlutterPlugin* instance = [[TDRemoteConfigFlutterPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
    
    FlutterEventChannel* eventChannel = [FlutterEventChannel
                                             eventChannelWithName:@"thinkingdata.cn/RemoteConfig/event"
                                             binaryMessenger:[registrar messenger]];
    [eventChannel setStreamHandler:instance];
    
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"enableLog" isEqualToString:call.method]) {
      bool enableLog = call.arguments;
      [TDRemoteConfig enableLog:enableLog];
      result(nil);
  }else if([@"initWithConfig" isEqualToString:call.method]){
      TDRemoteConfigSettings *settings = [[TDRemoteConfigSettings alloc]init];
      settings.appId = call.arguments[@"appId"];
      settings.serverUrl = call.arguments[@"serverUrl"];
      NSNumber *modeNumber = call.arguments[@"mode"];
      if(modeNumber && modeNumber.intValue == 1){
          settings.mode = TDRemoteConfigModeDebug;
      }
      settings.fetchParams = call.arguments[@"customFetchParams"];
      [TDRemoteConfig startWithSettings:settings];
      result(nil);
  }else if([@"setDefaultValues" isEqualToString:call.method]){
      [TDRemoteConfig setDefaultValues:call.arguments[@"defaultValues"] appId:call.arguments[@"appId"]];
      result(nil);
  }else if([@"setDefaultValuesWithFile" isEqualToString:call.method]){
      [TDRemoteConfig setDefaultValuesWithJsonFile:call.arguments[@"defaultValues"] appId:call.arguments[@"appId"]];
      result(nil);
  }else if([@"clearDefaultValues" isEqualToString:call.method]){
      [TDRemoteConfig clearDefaultValuesWithAppId:call.arguments[@"appId"]];
      result(nil);
  }else if([@"setCustomFetchParams" isEqualToString:call.method]){
      [TDRemoteConfig setCustomFetchParams:call.arguments[@"fetchParams"] appId:call.arguments[@"appId"]];
      result(nil);
  }else if([@"fetch" isEqualToString:call.method]){
      [TDRemoteConfig fetchWithAppId:call.arguments[@"appId"]];
      result(nil);
  }else if([@"removeCustomFetchParam" isEqualToString:call.method]){
      [TDRemoteConfig removeCustomFetchParam:call.arguments[@"key"] appId:call.arguments[@"appId"]];
      result(nil);
  }else if([@"addClientParams" isEqualToString:call.method]){
      [TDRemoteConfig addClientParams:call.arguments[@"clientParams"]];
      result(nil);
  }else if([@"getClientParams" isEqualToString:call.method]){
      NSDictionary *clientParams = [TDRemoteConfig getClientParams];
      result(clientParams);
  }else if([@"removeClientParam" isEqualToString:call.method]){
      [TDRemoteConfig removeClientParam:call.arguments[@"key"]];
      result(nil);
  }else if([@"accumulateNum" isEqualToString:call.method]){
      [TDRemoteConfig accumulateNum:call.arguments[@"num"] toClientParamKey:call.arguments[@"key"]];
      result(nil);
  }else if([@"getClientParamValueForKey" isEqualToString:call.method]){
      result([TDRemoteConfig getClientParamValueForKey:call.arguments[@"key"]]);
  }else if([@"getConfigValue" isEqualToString:call.method]){
      NSString *appId = call.arguments[@"appId"];
      NSArray<NSDictionary *> *dataList = call.arguments[@"dataList"];
      NSString *type = call.arguments[@"type"];
      if(dataList){
          TDObject *tdObject = [TDRemoteConfig getDataWithAppId:appId];
          for (NSDictionary *item in dataList) {
              if(!item) continue;
              id key = item[@"key"];
              id defaultValue = item[@"defaultValue"];
              if ([key isKindOfClass:[NSString class]]) {
                  tdObject = tdObject.getWithDefault((NSString *)key,defaultValue);
              } else if ([key isKindOfClass:[NSNumber class]]) {
                  NSNumber *num = (NSNumber *)key;
                  tdObject = tdObject[num.intValue];
              }
          }
          if ([@"string" isEqualToString:type]) {
              result([tdObject stringValue]);
              return;
          } else if ([@"int" isEqualToString:type]) {
              result([tdObject longNumber]);
              return;
          }else if ([@"double" isEqualToString:type]) {
              result([tdObject doubleNumber]);
              return;
          }else if ([@"bool" isEqualToString:type]) {
              NSNumber *numberValue = [tdObject numberValue];
              if (numberValue) {
                  result(numberValue);
              }
              result(@NO);
              return;
          }else if ([@"object" isEqualToString:type]) {
              result([tdObject objectValue]);
              return;
          }else if ([@"array" isEqualToString:type]) {
              result([tdObject arrayValue]);
              return;
          }
      }
      result(nil);
  }
  else if([@"addConfigFetchListener" isEqualToString:call.method]){
      [[NSNotificationCenter defaultCenter] addObserver:TDRemoteConfigReceiver.class selector:@selector(fun:) name:kTDRemoteConfigFetchDataSuccess object:nil];
      result(nil);
  }else {
    result(FlutterMethodNotImplemented);
  }
}
- (FlutterError* _Nullable)onListenWithArguments:(id _Nullable)arguments eventSink:(FlutterEventSink)events {
    [TDRemoteConfigReceiver setEventSink:events];
    return nil;
}

- (FlutterError* _Nullable)onCancelWithArguments:(id _Nullable)arguments {
    [TDRemoteConfigReceiver setEventSink:nil];
    return nil;
}

@end
